src/libpriv/rpmostree-core.c

```diff
@@ -1266,95 +1274,186 @@ sort_packages (RpmOstreeContext *self,
    * seem to be a way to tell libsolv to not touch some pkgs in the base layer,
    * so we just inspect its solution in retrospect. libdnf has the concept of
    * protected packages, but it still allows updating protected packages. */
 
+  HyGoal goal = dnf_context_get_goal (self->hifctx);
+
+  g_autoptr(GPtrArray) packages = NULL;
+
+  /* all strings are owned by pool */
+  g_autoptr(GPtrArray) forbidden = g_ptr_array_new ();
+
+  g_assert (!self->pkgs_to_remove);
+  self->pkgs_to_remove = g_hash_table_new_full (g_str_hash, g_str_equal, g_free,
+                                                (GDestroyNotify)g_variant_unref);
   packages = dnf_goal_get_packages (goal,
                                     DNF_PACKAGE_INFO_REMOVE,
-                                    DNF_PACKAGE_INFO_OBSOLETE,
-                                    -1);
-
-  /* strings are owned by pool */
-  g_autoptr(GPtrArray) filtered_packages = g_ptr_array_new ();
+                                    DNF_PACKAGE_INFO_OBSOLETE, -1);
   for (guint i = 0; i < packages->len; i++)
     {
       DnfPackage *pkg = packages->pdata[i];
       const char *name = dnf_package_get_name (pkg);
-      if (g_strv_contains (allowed_base_removals, name))
-        continue;
+      const char *nevra = dnf_package_get_nevra (pkg);
 
-      g_ptr_array_add (filtered_packages, (gpointer)dnf_package_get_nevra (pkg));
+      /* did we expect this package to be removed? */
+      if (rpmostree_str_ptrarray_contains (removed_pkgnames, name))
+        g_hash_table_insert (self->pkgs_to_remove, g_strdup (name), vnevra_from_pkg (pkg));
+      else
+        g_ptr_array_add (forbidden, (gpointer)nevra);
     }
 
-  if (filtered_packages->len > 0)
+  if (forbidden->len > 0)
+    return throw_package_list (error, "marked to be removed", forbidden);
+
+  g_ptr_array_unref (forbidden);
+  forbidden = g_ptr_array_new ();
+
+  /* check that all the pkgs we expect to remove are marked for removal */
+  GLNX_HASH_TABLE_FOREACH_KV (self->pkgs_to_remove, const char*, name, GVariant*, vnevra)
     {
-      g_autofree char *msg = join_package_list (
-          "The following base packages would be removed: ",
-          (char**)filtered_packages->pdata, filtered_packages->len);
-      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED, msg);
-      return FALSE;
+      if (!vnevra)
+        g_ptr_array_add (forbidden, (gpointer)name);
     }
 
+  if (forbidden->len > 0)
+    return throw_package_list (error, "not marked to be removed", forbidden);
+
+  /* REINSTALLs should never happen since it doesn't make sense in the rpm-ostree flow, and
+   * we check very early whether a package is already in the rootfs or not, but let's check
+   * for it anyway so that we get a bug report in case it somehow happens. */
   g_ptr_array_unref (packages);
+  packages = dnf_goal_get_packages (goal, DNF_PACKAGE_INFO_REINSTALL, -1);
+  g_assert (packages->len == 0);
```

Comment:
`g_assert_cmpint` in general so we get the values printed.  Although in this case it's probably worth an explicit error with the package strings.  (But I'm fine leaving it with just `g_assert_cmpint()` too).